pkg/updateengine/client.go

```diff
@@ -76,56 +83,175 @@ func dbusMember(iface, member string) string {
-	// move onto the main loop.
-	st, err := c.GetStatus()
+const (
+	dbusSdName      = "org.freedesktop.systemd1"
+	dbusSdManager   = "/org/freedesktop/systemd1"
+	dbusSdManagerIf = "org.freedesktop.systemd1.Manager"
+)
+
+func (c *Client) disableRPMOSTreeAutoUpdates() error {
+	// for now, just mask the systemd service -- alternatively, also mount
+	// /etc/rpm-ostreed.conf and turn off setting permanently?
+	mgr := c.conn.Object(dbusSdName, dbus.ObjectPath(dbusSdManager))
+
+	err := mgr.Call(dbusMember(dbusSdManagerIf, "MaskUnitFiles"), 0,
+	                [1]string{"rpm-ostreed-automatic.service"}, false, true).Err
 	if err != nil {
-		glog.Warningf("error getting rpm-ostree status: %v", err)
-	} else {
-		rcvr <- st
+		return err
 	}
+
+	err = mgr.Call(dbusMember(dbusSdManagerIf, "Reload"), 0).Err
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
-// ReceiveStatuses receives signal messages from dbus and sends them as Statuses
+// WatchForUpdate regularly checks for new rpm-ostree updates receives signal messages from dbus and sends them as Statuses
 // on the rcvr channel, until the stop channel is closed. An attempt is made to
 // get the initial status and send it on the rcvr channel before receiving
 // starts.
-func (c *Client) ReceiveStatuses(rcvr chan Status, stop <-chan struct{}) {
-	c.sendStatusUpdate(rcvr)
-	// XXX: We could probably increase the interval even more if we use
-	// filesystem notifications instead if we want to narrow the window between
-	// deployment staging and reboot.
+func (c *Client) WatchForUpdate(rcvr chan Status, stop <-chan struct{}) {
 	t := time.Tick(statePollInterval)
 	for {
 		select {
 		case <-stop:
 			return
 		case <-t:
-			c.sendStatusUpdate(rcvr)
+			c.checkForUpdate(rcvr)
 		}
 	}
 }
 
-// GetStatus gets the current status from rpm-ostree
-func (c *Client) GetStatus() (Status, error) {
-	glog.Info("Getting status from rpm-ostree")
+// Check for an update, and stages the update if found.
+func (c *Client) checkForUpdate(rcvr chan Status) {
+	glog.Info("Checking for update")
+
+	// We could go straight to Upgrade() here. The nice thing with using
+	// AutomaticUpdateTrigger() first is to make it easier eventually to split
+	// out "checking" from "downloading & staging".
+	rcvr <- Status{CurrentStatus: RpmOstreeUpdateChecking}
 
 	bootedOS := c.conn.Object(dbusName, c.osPath)
-	prop, err := bootedOS.GetProperty(dbusMember(dbusIfaceOS, "CachedUpdate"))
+	err := callAutomaticUpdateTrigger(bootedOS)
+	if err != nil {
+		glog.Warningf("Error calling AutomaticUpdateTrigger(): %v", err)
+		rcvr <- NewStatus(RpmOstreeUpdateError, nil)
+		return
+	}
+
+	cachedUpdate, err := getCachedUpdate(bootedOS)
+	if err != nil {
+		glog.Warningf("Error reading CachedUpdate property: %v", err)
+		rcvr <- NewStatus(RpmOstreeUpdateError, nil)
+		return
+	}
+
+	if len(cachedUpdate) == 0 {
+		// No updates, we're done
+		rcvr <- NewStatus(RpmOstreeUpdateNone, nil)
+		return
+	}
+
+	err = callUpgrade(bootedOS)
```

Comment:
Why not just callUpgrade first, and then look at the cached update data after?

Reply:
Currently, `Upgrade()` doesn't actually update `CachedUpdate`. That's part of https://github.com/projectatomic/rpm-ostree/pull/1321, though we can probably split that out. Though the reasoning overall for doing a check first and upgrade after is mentioned higher up :arrow_up: 

```
// We could go straight to Upgrade() here. The nice thing with using
// AutomaticUpdateTrigger() first is to make it easier eventually to split
// out "checking" from "downloading & staging".
```

I initially did have an `RpmOstreeUpdateAvailable` status which motivated this more. Though I removed it since there isn't really any need for it yet. I'm open to simplifying this back down to just calling `Upgrade()` and then post facto checking if a new deployment was actually laid down like we do in rpm-ostree.